server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -763,21 +751,78 @@ public void onFailure(Exception e) {
                         });
 
                         IngestDocument ingestDocument = newIngestDocument(indexRequest);
-                        executePipelines(pipelines.iterator(), hasFinalPipeline, indexRequest, ingestDocument, documentListener);
-
+                        LinkedHashSet<String> indexRecursionDetection = new LinkedHashSet<>();
+                        indexRecursionDetection.add(indexRequest.index());
```

Comment:
More general question: I assume during the routing, `index()` will always be the data_stream name for data_streams and the data_stream is only resolved to a write index after the processing happened? Asking it here because the data stream name is the one that should be part of the index recursion detection, not the write index as this could even change during ingest time I assume.

Reply:
Yes, _index isn't the internal write index of the data stream.